Add docker spinup module for Flask

blazebissar opened the pull request:
Checkout the [Engineer Workflow Doc ](https://docs.google.com/presentation/d/1pdzuHG-yO7_qW6jiByaGTLztKZ6b72lBNqSdNT5WZQ8/edit)to make sure you're contributing correctly! :)

**[Tickets](https://trello.com/b/dYOQKJGv/reclass)**

https://trello.com/c/0CWxFvSi

Additional Notes: Part of the backend [infrastructure development plan](https://docs.google.com/presentation/d/1a-DAE6mEX_oD6gT5UwZsz1r5OdsObwYBSxjLp1i2dT8/edit#slide=id.gb3f99ac2df_0_473) not captured in tickets.

**[Dependencies](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.czpk4cbi7z4g)**
Catpured in the [design document](https://docs.google.com/presentation/d/1a-DAE6mEX_oD6gT5UwZsz1r5OdsObwYBSxjLp1i2dT8/edit#slide=id.gb3f99ac2df_0_473) not captured in tickets.

**[Explain Impact on DB](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.285lgcs4z84o)**
None.

**[Testing](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.rs692kr58cf8)**

How did you test this? List of actions & expected behavior with helpful screenshots for reviewer:

Hello World example builds images, starts containers, checks resources, and kills started container. Unit testing suite runnable with `pytest -v` from the project root.

If this change is performance related, please document what tests were / should be done to verify improvement:
No performance changes.

takakonishimura commented on mininet_spinup.py:
```diff
@@ -0,0 +1,194 @@
+        run_mininet_container
+        build_mininet_container
+    Public static methods:
+        check_cpou_availability
```
minor typo `check_cpou_availability`

takakonishimura commented on mininet_spinup.py:
```diff
@@ -0,0 +1,194 @@
+        client: a DockerClient to communicate with the Docker daemon
+        dev_image: an Image object of the development Toynet image
+        prod_image: an Image object of the production Toynet image
+        runnint_containers: Dict(String: Container) containing all the started
```
minor typo `runnint_containers`

takakonishimura commented on mininet_spinup.py:
```diff
@@ -0,0 +1,194 @@
+        Builds the Toynet container from the local Dockerfile or dev.Dockerfile.
+
+        dev: bool - True indicates to build the dev.Dockerfile image, False
+                    indicates to build the production Dockerfile image
```
what do you think about enabling the user to specify the name of these files?

takakonishimura commented on mininet_spinup.py:
```diff
@@ -0,0 +1,194 @@
+        '''
+        Kills the specified Docker container if running.
+
+        conainer_name: String - the name of the container to kill
```
minor typo

takakonishimura commented on mininet_spinup.py:
```diff
@@ -0,0 +1,194 @@
+            del self.running_containers[container_name]
+        except (KeyError, docker.errors.APIError) as error:
+            print(error)
+            res = False
```
If we don't catch this exception, I believe we would return nothing, so it may make more sense to just return `True` at the end and not worry about `False`

takakonishimura commented on mininet_spinup.py:
```diff
@@ -0,0 +1,194 @@
+        return res
+
+    @staticmethod
+    def check_cpu_availability(thresholds=None):
```
let's accept variables and turn it into a list inside the container as the user doesn't need to know how this is done underneath the hood.

takakonishimura commented on mininet_spinup.py:
```diff
@@ -0,0 +1,194 @@
+    if test_toynet.check_cpu_availability():
+        print('Sufficient CPU')
+    else:
+        print('Indufficient CPU')
```
minor typo `Indufficient`

takakonishimura reviewed (changes requested):
Code looks good so far. Since this PR is supposed to check resources, it would be nice to have a documented way of hitting insufficient memory or cpu availability but I am not sure if there's something nice for this. Have you looked into anything like that?

takakonishimura commented on mininet_spinup.py:
```diff
@@ -0,0 +1,194 @@
+        else:
+            self.prod_image = self.client.images.build(path='./', rm=True, tag='toynet')
+
+    def kill_container(self, container_name):
```
should this be in `mininet_spindown.py` or should we rename the file?

blazebissar commented:
Thanks @takakonishimura, your comments should all be addressed now. Also added unit tests with an updated `README.md`.

takakonishimura commented on toynet_manager.py:
```diff
@@ -0,0 +1,218 @@
+        threshold: float - the maximum RAM usage above which we consider the
+                           system unable to support another Toynet container.
+        min_mem: int - the minimum number of GiB of available memory we
+                       requuire to consider the system able to support another
```
typo: `requuire` -> `require`

takakonishimura commented on toynet_manager.py:
```diff
@@ -0,0 +1,218 @@
+        '''
+
+        # CPU percent usage representation for the past 1, 5, and 15 minutes
+        loads = [load / psutil.cpu_count() * 100 for load in psutil.getloadavg()]
```
I'm not sure what the right answer is here, but https://docs.python.org/3/library/os.html#os.getloadavg appears to throw an OS error exception. https://psutil.readthedocs.io/en/latest/ does not. Should we expect that psutil library does not throw exceptions?

Additionally, in the second link the API says

>On Windows this is emulated by using a Windows API that spawns a thread which keeps running in background and updates results every 5 seconds, mimicking the UNIX behavior. Thus, on Windows, the first time this is called and for the next 5 seconds it will return a meaningless (0.0, 0.0, 0.0) tuple.

I suppose this means we should only ever run flask in unix? If we want to support windows it seems like we should call this function once somewhere in application startup and handle `0s`.

ajlacey replied:
@takakonishimura I don't think the Windows functionality will matter because we're not running toynet on Windows (it's on the container which is built from an Ubuntu image). 

If we had plans to run this on a Windows system it would need to be ran a minimum 5 seconds prior to show accurate non-zero information- but that info would be limited to the last 5 seconds. That is to say the windows behavior will not give accurate information on the 1/5/15 minute intervals until those intervals have passed from the initial run of getloadavg(). 

All of that said, it seems safe to assume that the resources _should_ be available when this is initially run.

blazebissar replied:
Found in the `psutil` [code](https://github.com/giampaolo/psutil/blob/6e494bd4024c3d25769688b53e95942e529392d6/psutil/__init__.py#L1898) that there is a 1:1 link to `os.getloadavg`, added error handling accordingly.

I don't think we intend to run our backend Flask server on Windows, do we? If so I can easily add to the `__init__` function (and even force a 5 second sleep before returning if we think setup latency is acceptable to guarantee good statistics on the first call).

takakonishimura commented on toynet_manager.py:
```diff
@@ -0,0 +1,218 @@
+        return True
+
+    @staticmethod
+    def check_memory_availability(threshold=75.0, min_mem=2):
```
How did you decide on 2 GB?

ajlacey replied:
Do we have any benchmarks on the resource utilization for an arbitrary container?

blazebissar replied:
2GB is the minimum memory recommended for Ubuntu, which is what I think we tend to run. 

A more minimal Debian install holds steady at 400-700GB of RAM usage, and even heavy Flask usage (with DB calls) does not spike us beyond an additional 30-40MB so this is more to keep the OS recommended resources available to minimize the chances of memory exhaustion. It seems reasonable to have a default that errs on the side of caution, unless you want us to be more deliberate?

takakonishimura reviewed (changes requested):
I like the unit tests. My understanding is that you will scale test this before finishing the mini-flask project (being able to report for some size VM how many containers we can create before the VM saturates) and you will also, at that time, address what happens between flask and  react when that limit is reached.

takakonishimura commented:
Can you make a dummy ticket assigned to you in `ToyNet Backend` and link this PR to it & put it `Ready for Review`?

blazebissar commented:
Addressed comments in code, can further modify depending on the default memory threshold discussion.

> I like the unit tests. My understanding is that you will scale test this before finishing the mini-flask project (being able to report for some size VM how many containers we can create before the VM saturates) and you will also, at that time, address what happens between flask and react when that limit is reached.

My understanding was that React and Flask were going to be on separate VMs? If they're going to be co-located then we can definitely accommodate. As it stands there would just be a returned error like 500 or 503 from Flask to React. But you are right that once we build this into Flask that we will confirm overall scalability/stress testing of the system, in particular because I would like to add in some testing that maintains high turnover of running containers to make sure resources work well.

ajlacey reviewed (approved):
Once the additional typo that tay caught is fixed, I think this should be good to merge.
